File: user-service/src/services/rabbitmq.producer.js
import amqp from "amqplib";

const RABBITMQ_URL = process.env.RABBITMQ_URL || "amqp://localhost";
const EXCHANGE_NAME = "social.events";

export class RabbitMQProducer {
    static connection = null;
    static channel = null;

    static async connect() {
        try {
            this.connection = await amqp.connect(RABBITMQ_URL);
            this.channel = await this.connection.createChannel();

            await this.channel.assertExchange(EXCHANGE_NAME, "topic", { durable: true });

            console.log("[RabbitMQ] Producer connected and exchange asserted.");

            this.connection.on("close", () => {
                console.error("[RabbitMQ] Connection closed. Retrying...");
                setTimeout(this.connect.bind(this), 5000);
            });

            this.connection.on("error", (err) => {
                console.error("[RabbitMQ] Association error", err);
            });

        } catch (error) {
            console.error("[RabbitMQ] Failed to connect", error);
            setTimeout(this.connect.bind(this), 5000);
        }
    }

    static async publish(routingKey, message) {
        if (!this.channel) {
            console.error("[RabbitMQ] Channel not active. Message dropped:", message);
            return;
        }

        try {
            this.channel.publish(
                EXCHANGE_NAME,
                routingKey,
                Buffer.from(JSON.stringify(message)),
                { persistent: true }
            );
            console.log(`[RabbitMQ] Published '${routingKey}'`);
        } catch (error) {
            console.error("[RabbitMQ] Failed to publish message", error);
        }
    }
}

File: post-service/rabbitmq_producer.py
import os
import json
import aio_pika

RABBITMQ_URL = os.getenv("RABBITMQ_URL", "amqp://rabbitmq")
EXCHANGE_NAME = "social.events"

async def publish_event(routing_key: str, message: dict):
    """
    Publish an event to RabbitMQ using the 'social.events' topic exchange.
    """
    try:
        connection = await aio_pika.connect_robust(RABBITMQ_URL)
        async with connection:
            channel = await connection.channel()
            # Declare topic exchange (same as notification-service consumer)
            exchange = await channel.declare_exchange(
                EXCHANGE_NAME,
                aio_pika.ExchangeType.TOPIC,
                durable=True
            )
            await exchange.publish(
                aio_pika.Message(
                    body=bytes(json.dumps(message), encoding="utf-8"),
                    delivery_mode=aio_pika.DeliveryMode.PERSISTENT
                ),
                routing_key=routing_key
            )
            print(f"[RabbitMQ] Published {routing_key}: {message}")
    except Exception as e:
        print(f"[RabbitMQ] Failed to publish {routing_key}: {str(e)}")
